src/NetAnalyzers/UnitTests/Microsoft.NetCore.Analyzers/Tasks/DoNotUseWhenAllOrWaitAllWithSingleArgumentTests.cs

```diff
@@ -0,0 +1,419 @@
+
+            ' Use WaitAll correctly with multiple tasks
+            Task.WaitAll(t1, t2)
+            Task.WaitAll(New Task() {t1, t2})
+            Task.WaitAll(CreateTask(), CreateTask())
+            Task.WaitAll(objectTask1, objectTask2)
+            Task.WaitAll(New Task() {objectTask1, objectTask2})
+            Task.WaitAll(CreateObjectTask(), CreateObjectTask())
+
+            ' Make sure a random Task.WaitAll that isn't System.Tasks.Task doens't trigger
+            Test.Task.WaitAll(t1)
+            Await Test.Task.WhenAll(t1)
+        End Function
+
+        Async Function WhenAllCall(ParamArray tasks As Task()) As Task
+            ' Should not trigger when taking an array from args
+            Await Task.WhenAll(tasks)
+        End Function
+
+        Sub WaitAllCall(ParamArray tasks As Task())
+            ' Should not trigger when taking an array from args
+            Task.WaitAll(tasks)
+        End Sub
+
+        Function CreateTask() As Task
+            Return Task.CompletedTask
+        End Function
+
+        Function CreateObjectTask() As Task(Of Object)
+            Return Task.FromResult(New Object())
+        End Function
+    End Class
+End Namespace
+
+Namespace Test
+    Public Module Task
+        Public Async Function WhenAll(ParamArray tasks As System.Threading.Tasks.Task()) As System.Threading.Tasks.Task
+            ' Should not trigger when taking an array from args
+            Await System.Threading.Tasks.Task.WhenAll(tasks)
+        End Function
+
+        Public Sub WaitAll(ParamArray tasks As System.Threading.Tasks.Task())
+            ' Should not trigger when taking an array from args
+            System.Threading.Tasks.Task.WaitAll(tasks)
+        End Sub
+    End Module
+End Namespace
+");
+        }
+
+        [Fact]
+        public async Task Diagnostics_FixApplies_WhenAll_CSharp()
+        {
+            var source = @"
+using System.Threading.Tasks;
+
+class C
+{
+    async Task M()
+    {
+        var t1 = CreateTask();
+        var objectTask1 = CreateObjectTask();
+
+        await {|CA1840:Task.WhenAll(t1)|};
+        await {|CA1840:Task.WhenAll(CreateTask())|};
+        var t1WhenAll = {|CA1840:Task.WhenAll(t1)|};
+        DoSomethingWithTask(t1WhenAll);
+
+        await {|CA1840:Task.WhenAll(objectTask1)|};
+        await {|CA1840:Task.WhenAll(CreateObjectTask())|};
+        var ot1WhenAll = {|CA1840:Task.WhenAll(objectTask1)|};
+        DoSomethingWithTask(ot1WhenAll);
+    }
+
+    void DoSomethingWithTask(Task task) 
+    {
+    }
+
+    Task CreateTask() => Task.CompletedTask;
+    Task<object> CreateObjectTask() => Task.FromResult(new object());
+}
+";
+
+            var fixedSource = @"
+using System.Threading.Tasks;
+
+class C
+{
+    async Task M()
+    {
+        var t1 = CreateTask();
+        var objectTask1 = CreateObjectTask();
+
+        await t1;
+        await CreateTask();
+        var t1WhenAll = {|CA1840:Task.WhenAll(t1)|};
```

Comment:
Probably also add the case where this is an assignment operation, not a variable declaration/initialization.